### DISCUSSION TITLE
Eye Square with Rounded Corner - Just a Question, not an Issue

### DISCUSSION THREAD

**Discussion by SergiuPogorDev** [Bacon/BaconQrCode#165] [Q&A]
I'm trying to write a new Module Renderer for Eye(internal and external) in order to get Square with Rounded Corner(not circle) but not sure how to get that. 

**I mean, using matrix logic how to achieve this?**

For Circle I use this one:
```
      ->move(3.5, 0)
      ->ellipticArc(3.5, 3.5, 0., false, true, 0., 3.5)
      ->ellipticArc(3.5, 3.5, 0., false, true, -3.5, 0.)
      ->ellipticArc(3.5, 3.5, 0., false, true, 0., -3.5)
      ->ellipticArc(3.5, 3.5, 0., false, true, 3.5, 0.)
      ->close()
```

```
      ->move(2.5, 0)
      ->ellipticArc(2.5, 2.5, 0., false, true, 0., 2.5)
      ->ellipticArc(2.5, 2.5, 0., false, true, -2.5, 0.)
      ->ellipticArc(2.5, 2.5, 0., false, true, 0., -2.5)
      ->ellipticArc(2.5, 2.5, 0., false, true, 2.5, 0.)
      ->close()
```

**Should I focus on replacing ellipticArc with a combination of line() and curve()?**

Expected result:
 ![qr-code](https://user-images.githubusercontent.com/79079906/194512254-d6beaa34-d237-475a-becb-d0ecc4eeaaf4.png)

**Comment by DASPRiD** [comment]
I'd recommend to build it in e.g. Inkscape and have a look at the SVG path code. You can basically do the exact same thing then in code here.

**Comment by enpesavento** [comment]
@SergiuPogor I did some experimenting and was able to create something similar using line() and curve().

I created a new renderer module implementing the EyeInterface and created the paths as follows:

**External Path**:
```
$outerSize = 3.5;
$outerMidPoint = 1.0;

$innerSize = 2.5;
$innerMidPoint = 1.0;

(new Path())
    ->move(0, $outerSize)
    ->line($outerMidPoint, $outerSize)
    ->curve($outerMidPoint, $outerSize, $outerSize, $outerSize, $outerSize, $outerMidPoint)
    ->line($outerSize, 0)
    ->line($outerSize, -$outerMidPoint)
    ->curve($outerSize, -$outerMidPoint, $outerSize, -$outerSize, $outerMidPoint, -$outerSize)
    ->line(0, -$outerSize)
    ->line(-$outerMidPoint, -$outerSize)
    ->curve(-$outerMidPoint, -$outerSize, -$outerSize, -$outerSize, -$outerSize, -$outerMidPoint)
    ->line(-$outerSize, 0)
    ->line(-$outerSize, $outerMidPoint)
    ->curve(-$outerSize, $outerMidPoint, -$outerSize, $outerSize, -$outerMidPoint, $outerSize)
    ->line(0, $outerSize)
    ->close()

    ->move(0, $innerSize)
    ->line($innerMidPoint, $innerSize)
    ->curve($innerMidPoint, $innerSize, $innerSize, $innerSize, $innerSize, $innerMidPoint)
    ->line($innerSize, 0)
    ->line($innerSize, -$innerMidPoint)
    ->curve($innerSize, -$innerMidPoint, $innerSize, -$innerSize, $innerMidPoint, -$innerSize)
    ->line(0, -$innerSize)
    ->line(-$innerMidPoint, -$innerSize)
    ->curve(-$innerMidPoint, -$innerSize, -$innerSize, -$innerSize, -$innerSize, -$innerMidPoint)
    ->line(-$innerSize, 0)
    ->line(-$innerSize, $innerMidPoint)
    ->curve(-$innerSize, $innerMidPoint, -$innerSize, $innerSize, -$innerMidPoint, $innerSize)
    ->line(0, $innerSize)
    ->close();
```

**Internal Path**:
```
$size = 1.5;
$midPoint = 0.75;

(new Path())
    ->move(0, $size)
    ->line($midPoint, $size)
    ->curve($midPoint, $size, $size, $size, $size, $midPoint)
    ->line($size, 0)
    ->line($size, -$midPoint)
    ->curve($size, -$midPoint, $size, -$size, $midPoint, -$size)
    ->line(0, -$size)
    ->line(-$midPoint, -$size)
    ->curve(-$midPoint, -$size, -$size, -$size, -$size, -$midPoint)
    ->line(-$size, 0)
    ->line(-$size, $midPoint)
    ->curve(-$size, $midPoint, -$size, $size, -$midPoint, $size)
    ->line(0, $size)
    ->close();
```

Here the result:
![test](https://user-images.githubusercontent.com/25293998/210568988-7164584b-8fec-4107-a9bf-5b956e7437e3.svg)

**Comment by wojoj93** [comment]
Could someone explain to me why using this code I get very poor quality rounding?
![Zaznaczenie_1204](https://github.com/Bacon/BaconQrCode/assets/18478789/bf90a079-3550-46c0-ba15-e8b1469fc5f3)

**Comment by Rizwansayyed786** [comment]
how can I use this to work in mine I would want to build the same QR code. I am using Laravel?

this is my current generated qr code using simple qr code
![fonp-qr](https://github.com/Bacon/BaconQrCode/assets/70570006/e0d72165-2261-40c3-ba75-541c4e565b48)

**Comment by IggsGrey** [comment]
@wojoj93 that's probably a png you're rendering?

Try it in svg format, it's much smoother.

The png renderer probably can't handle the curves properly for whatever reason.

**Comment by glossyweb** [comment]
![Screenshot_71](https://github.com/user-attachments/assets/de795379-2364-46d6-a0b4-e0cfa1dca14b)
Somebody know how to draw same inner and outer shapes?
